Tests sometimes cannot spin up an internal IPFS node

We have a loop where we attempt to spin up an in-process IPFS node using a random port, but this is implemented as a hack. We should find a way that doesn't involve looping  a number of times hoping that one works.

https://github.com/bacalhau-project/bacalhau/blob/a75a7bf696bb9a2d70b394b6b5f02bf8119c3dc7/pkg/ipfs/node.go#L138

Example test error:

```
15:53:01.415 | ERR ../../../../go/pkg/mod/github.com/ipfs/kubo@v0.18.2-0.20230214211946-0ff406170dd9/core/builder.go:158 > constructing the node: failed to listen on any addresses: [listen tcp4 0.0.0.0:37557: bind: address already in use] [logger-name:core]
    storage_test.go:29: 
        	Error Trace:	/home/circleci/repo/pkg/storage/ipfs/storage_test.go:29
        	            				/home/circleci/repo/pkg/storage/ipfs/storage_test.go:70
        	Error:      	Received unexpected error:
        	            	failed to create ipfs node: failed to create node: constructing the node (see log for full detail): failed to listen on any addresses: [listen tcp4 0.0.0.0:37557: bind: address already in use]
        	Test:       	TestPrepareStorageRespectsTimeouts/1m0s
    --- FAIL: TestPrepareStorageRespectsTimeouts/1m0s (0.49s)
```
https://app.circleci.com/pipelines/github/bacalhau-project/bacalhau/4463/workflows/cc05ce61-7eb8-46d1-b752-43b5088e9ba9/jobs/57123

--- a/pkg/devstack/devstack.go
+++ b/pkg/devstack/devstack.go
@@ -8,18 +8,18 @@ import (
 	"strings"
 	"time"
 
+	"github.com/bacalhau-project/bacalhau/pkg/config"
+	"github.com/bacalhau-project/bacalhau/pkg/ipfs"
 	"github.com/bacalhau-project/bacalhau/pkg/jobstore/inmemory"
 	"github.com/bacalhau-project/bacalhau/pkg/libp2p"
 	"github.com/bacalhau-project/bacalhau/pkg/logger"
-	filecoinlotus "github.com/bacalhau-project/bacalhau/pkg/publisher/filecoin_lotus"
-	"github.com/imdario/mergo"
-	"github.com/libp2p/go-libp2p/core/host"
-
-	"github.com/bacalhau-project/bacalhau/pkg/config"
-	"github.com/bacalhau-project/bacalhau/pkg/ipfs"
 	"github.com/bacalhau-project/bacalhau/pkg/model"
 	"github.com/bacalhau-project/bacalhau/pkg/node"
+	filecoinlotus "github.com/bacalhau-project/bacalhau/pkg/publisher/filecoin_lotus"
 	"github.com/bacalhau-project/bacalhau/pkg/system"
+	"github.com/bacalhau-project/bacalhau/pkg/util/multiaddresses"
+	"github.com/imdario/mergo"
+	"github.com/libp2p/go-libp2p/core/host"
 	"github.com/multiformats/go-multiaddr"
 	"github.com/phayes/freeport"
 	"github.com/rs/zerolog/log"
@@ -201,9 +201,10 @@ func NewDevStack(
 			if err != nil {
 				return nil, err
 			}
+			addresses := multiaddresses.SortLocalhostFirst(nodes[0].Host.Addrs())
 			// Only use a single address as libp2p seems to have concurrency issues, like two nodes not able to finish
 			// connecting/joining topics, when using multiple addresses for a single host.
-			libp2pPeer = append(libp2pPeer, nodes[0].Host.Addrs()[0].Encapsulate(p2pAddr))
+			libp2pPeer = append(libp2pPeer, addresses[0].Encapsulate(p2pAddr))
 			log.Ctx(ctx).Debug().Msgf("Connecting to first libp2p requester node: %s", libp2pPeer)
 		}
 

--- a/pkg/ipfs/address_in_use_other.go
+++ b/pkg/ipfs/address_in_use_other.go
@@ -1,9 +1,0 @@
-//go:build !unix
-
-package ipfs
-
-import (
-	"errors"
-)
-
-var addressInUseError = errors.New("unsure what error other OSes give for this so have an error which won't match anything")

--- a/pkg/ipfs/address_in_use_unix.go
+++ b/pkg/ipfs/address_in_use_unix.go
@@ -1,7 +1,0 @@
-//go:build unix
-
-package ipfs
-
-import "syscall"
-
-var addressInUseError = syscall.EADDRINUSE

--- a/pkg/ipfs/client.go
+++ b/pkg/ipfs/client.go
@@ -5,10 +5,10 @@ import (
 	"fmt"
 	"net/http"
 	"os"
-	"sort"
 
 	"github.com/bacalhau-project/bacalhau/pkg/system"
 	"github.com/bacalhau-project/bacalhau/pkg/util/generic"
+	"github.com/bacalhau-project/bacalhau/pkg/util/multiaddresses"
 	httpapi "github.com/ipfs/go-ipfs-http-client"
 	ipld "github.com/ipfs/go-ipld-format"
 	files "github.com/ipfs/go-libipfs/files"
@@ -129,24 +129,7 @@ func (cl Client) SwarmAddresses(ctx context.Context) ([]string, error) {
 	// This sorts the addresses so IPv4 localhost is first, with the aim of using the localhost connection during tests
 	// and so avoid any unneeded network hops. Other callers to this either sort the list themselves or just output the
 	// full list.
-	preferLocalhost := func(m ma.Multiaddr) int {
-		count := 0
-		if _, err := m.ValueForProtocol(ma.P_TCP); err == nil {
-			count++
-		}
-		if ip, err := m.ValueForProtocol(ma.P_IP4); err == nil {
-			count++
-			if ip == "127.0.0.1" {
-				count++
-			}
-		} else if ip, err := m.ValueForProtocol(ma.P_IP6); err == nil && ip != "::1" {
-			count++
-		}
-		return count
-	}
-	sort.Slice(multiAddresses, func(i, j int) bool {
-		return preferLocalhost(multiAddresses[i]) > preferLocalhost(multiAddresses[j])
-	})
+	multiAddresses = multiaddresses.SortLocalhostFirst(multiAddresses)
 
 	addresses := generic.Map(multiAddresses, func(f ma.Multiaddr) string {
 		return f.String()

--- a/pkg/ipfs/node.go
+++ b/pkg/ipfs/node.go
@@ -15,7 +15,6 @@ import (
 	"github.com/libp2p/go-libp2p/core/peer"
 	ma "github.com/multiformats/go-multiaddr"
 	manet "github.com/multiformats/go-multiaddr/net"
-	"github.com/phayes/freeport"
 	"github.com/pkg/errors"
 
 	"github.com/rs/zerolog/log"
@@ -59,8 +58,7 @@ type Node struct {
 	// APIPort is the port that the node's ipfs API is listening on.
 	APIPort int
 
-	// SwarmPort is the port that the node's ipfs swarm is listening on.
-	SwarmPort int
+	apiAddresses []string
 }
 
 // NodeMode configures how the node treats the public IPFS network.
@@ -104,10 +102,6 @@ func (cfg *Config) getMode() NodeMode {
 	return cfg.Mode
 }
 
-func (cfg *Config) getPeerAddrs() []string {
-	return cfg.PeerAddrs
-}
-
 // NewNode creates a new IPFS node in default mode, which creates an IPFS
 // repo in a temporary directory, uses the public libp2p nodes as peers and
 // generates a repo keypair with 2048 bits.
@@ -129,31 +123,10 @@ func newNode(ctx context.Context, cm *system.CleanupManager, peerAddrs []string,
 			filteredPeerAddrs = append(filteredPeerAddrs, addr)
 		}
 	}
-	return tryCreateNode(ctx, cm, Config{
+	return newNodeWithConfig(ctx, cm, Config{
 		Mode:      mode,
 		PeerAddrs: filteredPeerAddrs,
 	})
-}
-
-func tryCreateNode(ctx context.Context, cm *system.CleanupManager, cfg Config) (*Node, error) {
-	// Starting up an IPFS node can have issues as there's a race between finding a free port and getting the listener
-	// running on that port (e.g. find the port, write the config file, save the file, start up IPFS, then start the listener)
-	attempts := 3
-	var err error
-	for i := 0; i < attempts; i++ {
-		var ipfsNode *Node
-		ipfsNode, err = newNodeWithConfig(ctx, cm, cfg)
-		if err != nil {
-			if errors.Is(err, addressInUseError) {
-				log.Ctx(ctx).Debug().Err(err).Msg("Failed to start up node as port was already in use")
-				continue
-			}
-			return nil, err
-		}
-
-		return ipfsNode, nil
-	}
-	return nil, err
 }
 
 // newNodeWithConfig creates a new IPFS node with the given configuration.
@@ -177,43 +150,29 @@ func newNodeWithConfig(ctx context.Context, cm *system.CleanupManager, cfg Confi
 		}
 	}()
 
-	if err = connectToPeers(ctx, api, ipfsNode, cfg.getPeerAddrs()); err != nil {
+	if err = connectToPeers(ctx, api, ipfsNode, cfg.PeerAddrs); err != nil {
 		log.Ctx(ctx).Error().Msgf("ipfs node failed to connect to peers: %s", err)
 	}
 
-	if err = serveAPI(cm, ipfsNode, repoPath); err != nil {
+	apiAddresses, err := serveAPI(cm, ipfsNode, repoPath)
+	if err != nil {
 		return nil, fmt.Errorf("failed to serve API: %w", err)
 	}
 
-	// Fetch useful info from the newly initialized node:
-	nodeCfg, err := ipfsNode.Repo.Config()
-	if err != nil {
-		return nil, fmt.Errorf("failed to get repo config: %w", err)
-	}
-
 	var apiPort int
-	if len(nodeCfg.Addresses.API) > 0 {
-		apiPort, err = getTCPPort(nodeCfg.Addresses.API[0])
+	if len(apiAddresses) > 0 {
+		apiPort, err = getTCPPort(apiAddresses[0])
 		if err != nil {
 			return nil, fmt.Errorf("failed to parse api port: %w", err)
 		}
 	}
 
-	var swarmPort int
-	if len(nodeCfg.Addresses.Swarm) > 0 {
-		swarmPort, err = getTCPPort(nodeCfg.Addresses.Swarm[0])
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse swarm port: %w", err)
-		}
-	}
-
 	n := Node{
-		api:       api,
-		ipfsNode:  ipfsNode,
-		Mode:      cfg.getMode(),
-		RepoPath:  repoPath,
-		APIPort:   apiPort,
-		SwarmPort: swarmPort,
+		api:      api,
+		ipfsNode: ipfsNode,
+		Mode:     cfg.getMode(),
+		RepoPath: repoPath,
+		APIPort:  apiPort,
 	}
 
 	cm.RegisterCallbackWithContext(n.Close)
@@ -230,30 +189,15 @@ func (n *Node) ID() string {
 	return n.ipfsNode.Identity.String()
 }
 
-// APIAddresses returns the node's api addresses.
-func (n *Node) APIAddresses() ([]string, error) {
-	cfg, err := n.ipfsNode.Repo.Config()
-	if err != nil {
-		return nil, fmt.Errorf("failed to get repo config: %w", err)
-	}
-
-	var res []string
-	for _, addr := range cfg.Addresses.API {
-		res = append(res, fmt.Sprintf("%s/p2p/%s", addr, n.ID()))
-	}
-
-	return res, nil
-}
-
 // SwarmAddresses returns the node's swarm addresses.
 func (n *Node) SwarmAddresses() ([]string, error) {
-	cfg, err := n.ipfsNode.Repo.Config()
+	addresses, err := n.api.Swarm().ListenAddrs(context.TODO())
 	if err != nil {
 		return nil, fmt.Errorf("failed to get repo config: %w", err)
 	}
 
 	var res []string
-	for _, addr := range cfg.Addresses.Swarm {
+	for _, addr := range addresses {
 		res = append(res, fmt.Sprintf("%s/p2p/%s", addr, n.ID()))
 	}
 
@@ -262,25 +206,15 @@ func (n *Node) SwarmAddresses() ([]string, error) {
 
 // LogDetails logs connection details for the node's swarm and API servers.
 func (n *Node) LogDetails() {
-	apiAddrs, err := n.APIAddresses()
-	if err != nil {
-		log.Debug().Msgf("error fetching api addresses: %s", err)
-		return
-	}
+	id := n.ID()
 
-	var swarmAddrs []string
-	swarmAddrs, err = n.SwarmAddresses()
+	swarmAddrs, err := n.SwarmAddresses()
 	if err != nil {
 		log.Debug().Msgf("error fetching swarm addresses: %s", err)
+	} else {
+		log.Trace().Str("id", id).Strs("addresses", swarmAddrs).Msg("IPFS node listening for swarm")
 	}
-
-	id := n.ID()
-	for _, apiAddr := range apiAddrs {
-		log.Trace().Msgf("IPFS node %s listening for API on: %s", id, apiAddr)
-	}
-	for _, swarmAddr := range swarmAddrs {
-		log.Trace().Msgf("IPFS node %s listening for swarm on: %s", id, swarmAddr)
-	}
+	log.Trace().Str("id", id).Strs("addresses", n.apiAddresses).Msg("IPFS node listening for API")
 }
 
 // Client returns an API client for interacting with the node.
@@ -346,22 +280,22 @@ func createNode(ctx context.Context, _ *system.CleanupManager, cfg Config) (icor
 }
 
 // serveAPI starts a new API server for the node on the given address.
-func serveAPI(cm *system.CleanupManager, node *core.IpfsNode, repoPath string) error {
+func serveAPI(cm *system.CleanupManager, node *core.IpfsNode, repoPath string) ([]string, error) {
 	cfg, err := node.Repo.Config()
 	if err != nil {
-		return fmt.Errorf("failed to get repo config: %w", err)
+		return nil, fmt.Errorf("failed to get repo config: %w", err)
 	}
 
 	var listeners []manet.Listener
 	for _, addr := range cfg.Addresses.API {
 		maddr, err := ma.NewMultiaddr(addr)
 		if err != nil {
-			return fmt.Errorf("failed to parse multiaddr: %w", err)
+			return nil, fmt.Errorf("failed to parse multiaddr: %w", err)
 		}
 
 		listener, err := manet.Listen(maddr)
 		if err != nil {
-			return fmt.Errorf("failed to listen on api multiaddr: %w", err)
+			return nil, fmt.Errorf("failed to listen on api multiaddr: %w", err)
 		}
 
 		cm.RegisterCallback(func() error {
@@ -392,16 +326,18 @@ func serveAPI(cm *system.CleanupManager, node *core.IpfsNode, repoPath string) e
 		corehttp.CommandsOption(cmdContext),
 	}
 
+	var addresses []string
 	for _, listener := range listeners {
+		addresses = append(addresses, listener.Multiaddr().String())
 		// NOTE: this is not critical, but we should log for debugging
 		go func(listener manet.Listener) {
 			if err := corehttp.Serve(node, manet.NetListener(listener), opts...); err != nil {
-				log.Debug().Msgf("node '%s' failed to serve ipfs api: %s", node.Identity, err)
+				log.Debug().Err(err).Msgf("node '%s' failed to serve ipfs api", node.Identity)
 			}
 		}(listener)
 	}
 
-	return nil
+	return addresses, nil
 }
 
 // connectToPeers connects the node to a list of IPFS bootstrap peers.
@@ -425,9 +361,7 @@ func connectToPeers(ctx context.Context, api icore.CoreAPI, node *core.IpfsNode,
 			defer wg.Done()
 			if err := api.Swarm().Connect(ctx, peerInfo); err != nil {
 				anyErr = err
-				log.Ctx(ctx).Debug().Msgf(
-					"failed to connect to ipfs peer %s, skipping: %s",
-					peerInfo.ID, err)
+				log.Ctx(ctx).Debug().Err(err).Msgf("failed to connect to ipfs peer %s, skipping", peerInfo.ID)
 			}
 		}(peerInfo)
 	}
@@ -456,27 +390,9 @@ func createRepo(path string, nodeConfig Config) error {
 		return err
 	}
 
-	var apiPort int
-	apiPort, err = freeport.GetFreePort()
-	if err != nil {
-		return fmt.Errorf("could not create port for api: %w", err)
-	}
-
 	// If we're in local mode, then we need to manually change the config to
 	// serve an IPFS swarm client on some local port:
 	if nodeConfig.getMode() == ModeLocal {
-		var gatewayPort int
-		gatewayPort, err = freeport.GetFreePort()
-		if err != nil {
-			return fmt.Errorf("could not create port for gateway: %w", err)
-		}
-
-		var swarmPort int
-		swarmPort, err = freeport.GetFreePort()
-		if err != nil {
-			return fmt.Errorf("could not create port for swarm: %w", err)
-		}
-
 		cfg.AutoNAT.ServiceMode = config.AutoNATServiceDisabled
 		cfg.Swarm.EnableHolePunching = config.False
 		cfg.Swarm.DisableNatPortMap = true
@@ -484,25 +400,17 @@ func createRepo(path string, nodeConfig Config) error {
 		cfg.Swarm.RelayService.Enabled = config.False
 		cfg.Swarm.Transports.Network.Relay = config.False
 		cfg.Discovery.MDNS.Enabled = false
-		cfg.Addresses.Gateway = []string{
-			fmt.Sprintf("/ip4/0.0.0.0/tcp/%d", gatewayPort),
-		}
-		cfg.Addresses.API = []string{
-			fmt.Sprintf("/ip4/0.0.0.0/tcp/%d", apiPort),
-		}
-		cfg.Addresses.Swarm = []string{
-			fmt.Sprintf("/ip4/0.0.0.0/tcp/%d", swarmPort),
-		}
+		cfg.Addresses.Gateway = []string{"/ip4/0.0.0.0/tcp/0"}
+		cfg.Addresses.API = []string{"/ip4/0.0.0.0/tcp/0"}
+		cfg.Addresses.Swarm = []string{"/ip4/0.0.0.0/tcp/0"}
 	} else {
-		cfg.Addresses.API = []string{
-			fmt.Sprintf("/ip4/127.0.0.1/tcp/%d", apiPort),
-		}
+		cfg.Addresses.API = []string{"/ip4/127.0.0.1/tcp/0"}
 	}
 
 	// establish peering with the passed nodes. This is different than bootstrapping or manually connecting to peers,
 	//and kubo will create sticky connections with these nodes and reconnect if the connection is lost
 	// https://github.com/ipfs/kubo/blob/master/docs/config.md#peering
-	swarmPeers, err := ParsePeersString(nodeConfig.getPeerAddrs())
+	swarmPeers, err := ParsePeersString(nodeConfig.PeerAddrs)
 	if err != nil {
 		return fmt.Errorf("failed to parse peer addresses: %w", err)
 	}

--- a/pkg/ipfs/node_test.go
+++ b/pkg/ipfs/node_test.go
@@ -83,9 +83,11 @@ func (s *NodeSuite) TestFunctionality() {
 	s.Require().Equal(testString, string(data))
 
 	s.Never(func() bool {
-		peers, err := n2.Client().API.Swarm().Peers(ctx)
+		peers, err := n2.Client().API.Swarm().KnownAddrs(ctx)
 		s.Require().NoError(err)
-
+		id, err := n2.Client().API.Key().Self(ctx)
+		s.Require().NoError(err)
+		delete(peers, id.ID())
 		return !s.Len(peers, 1)
 	}, 500*time.Millisecond, 10*time.Millisecond, "a local node should only connect to the passed in peers")
 

--- a/pkg/libp2p/host.go
+++ b/pkg/libp2p/host.go
@@ -29,23 +29,15 @@ func NewHost(port int, opts ...libp2p.Option) (host.Host, error) {
 	}
 
 	addrs := []string{
-		"/ip4/0.0.0.0/tcp/%d",
-		"/ip4/0.0.0.0/udp/%d/quic",
-		"/ip4/0.0.0.0/udp/%d/quic-v1",
-		"/ip6/::/tcp/%d",
-		"/ip6/::/udp/%d/quic",
-		"/ip6/::/udp/%d/quic-v1",
-	}
-	listenAddrs := make([]multiaddr.Multiaddr, 0, len(addrs))
-	for _, s := range addrs {
-		addr, addrErr := multiaddr.NewMultiaddr(fmt.Sprintf(s, port))
-		if addrErr != nil {
-			return nil, addrErr
-		}
-		listenAddrs = append(listenAddrs, addr)
+		fmt.Sprintf("/ip4/0.0.0.0/tcp/%d", port),
+		fmt.Sprintf("/ip4/0.0.0.0/udp/%d/quic", port),
+		fmt.Sprintf("/ip4/0.0.0.0/udp/%d/quic-v1", port),
+		fmt.Sprintf("/ip6/::/tcp/%d", port),
+		fmt.Sprintf("/ip6/::/udp/%d/quic", port),
+		fmt.Sprintf("/ip6/::/udp/%d/quic-v1", port),
 	}
 
-	opts = append(opts, libp2p.ListenAddrs(listenAddrs...))
+	opts = append(opts, libp2p.ListenAddrStrings(addrs...))
 	opts = append(opts, libp2p.Identity(prvKey))
 	h, err := libp2p.New(opts...)
 	if err != nil {

--- a/pkg/logger/logger_test.go
+++ b/pkg/logger/logger_test.go
@@ -5,7 +5,6 @@ package logger
 import (
 	"context"
 	"errors"
-	"fmt"
 	"io"
 	"os"
 	"regexp"
@@ -20,7 +19,6 @@ import (
 	"github.com/ipfs/kubo/plugin/loader"
 	kuboRepo "github.com/ipfs/kubo/repo"
 	"github.com/ipfs/kubo/repo/fsrepo"
-	"github.com/phayes/freeport"
 	"github.com/rs/zerolog"
 	"github.com/rs/zerolog/log"
 	"github.com/stretchr/testify/assert"
@@ -117,13 +115,6 @@ func triggerIPFSLogging(t *testing.T) {
 	require.NoError(t, err)
 	require.NoError(t, config.Profiles["test"].Transform(cfg))
 
-	apiPort, err := freeport.GetFreePort()
-	require.NoError(t, err)
-	gatewayPort, err := freeport.GetFreePort()
-	require.NoError(t, err)
-	swarmPort, err := freeport.GetFreePort()
-	require.NoError(t, err)
-
 	cfg.AutoNAT.ServiceMode = config.AutoNATServiceDisabled
 	cfg.Swarm.EnableHolePunching = config.False
 	cfg.Swarm.DisableNatPortMap = true
@@ -131,15 +122,9 @@ func triggerIPFSLogging(t *testing.T) {
 	cfg.Swarm.RelayService.Enabled = config.False
 	cfg.Swarm.Transports.Network.Relay = config.False
 	cfg.Discovery.MDNS.Enabled = false
-	cfg.Addresses.Gateway = []string{
-		fmt.Sprintf("/ip4/127.0.0.1/tcp/%d", gatewayPort),
-	}
-	cfg.Addresses.API = []string{
-		fmt.Sprintf("/ip4/127.0.0.1/tcp/%d", apiPort),
-	}
-	cfg.Addresses.Swarm = []string{
-		fmt.Sprintf("/ip4/127.0.0.1/tcp/%d", swarmPort),
-	}
+	cfg.Addresses.Gateway = []string{"/ip4/0.0.0.0/tcp/0"}
+	cfg.Addresses.API = []string{"/ip4/0.0.0.0/tcp/0"}
+	cfg.Addresses.Swarm = []string{"/ip4/0.0.0.0/tcp/0"}
 	cfg.Peering = config.Peering{
 		Peers: nil,
 	}

--- a/pkg/node/node.go
+++ b/pkg/node/node.go
@@ -82,13 +82,7 @@ type Node struct {
 }
 
 func (n *Node) Start(ctx context.Context) error {
-	go func(ctx context.Context) {
-		if err := n.APIServer.ListenAndServe(ctx, n.CleanupManager); err != nil {
-			log.Ctx(ctx).Error().Msgf("Api server can't run. Cannot serve client requests!: %v", err)
-		}
-	}(ctx)
-
-	return nil
+	return n.APIServer.ListenAndServe(ctx, n.CleanupManager)
 }
 
 func NewStandardNode(

--- a/pkg/publicapi/server.go
+++ b/pkg/publicapi/server.go
@@ -178,14 +178,17 @@ func (apiServer *APIServer) ListenAndServe(ctx context.Context, cm *system.Clean
 	// Cleanup resources when system is done:
 	cm.RegisterCallbackWithContext(srv.Shutdown)
 
-	err = srv.Serve(listener)
-	if err == http.ErrServerClosed {
-		log.Ctx(ctx).Debug().Msgf(
-			"API server closed for host %s on %s.", apiServer.Address, srv.Addr)
-		return nil // expected error if the server is shut down
-	}
+	go func() {
+		err := srv.Serve(listener)
+		if err == http.ErrServerClosed {
+			log.Ctx(ctx).Debug().Msgf(
+				"API server closed for host %s on %s.", apiServer.Address, srv.Addr)
+		} else if err != nil {
+			log.Ctx(ctx).Err(err).Msg("Api server can't run. Cannot serve client requests!")
+		}
+	}()
 
-	return err
+	return nil
 }
 
 func (apiServer *APIServer) RegisterHandlers(config ...HandlerConfig) error {

--- a/pkg/publicapi/util_test.go
+++ b/pkg/publicapi/util_test.go
@@ -15,21 +15,16 @@ import (
 const TimeToWaitForServerReply = 10
 const TimeToWaitForHealthy = 50
 
-//nolint:unused // used in tests
 func setupNodeForTest(t *testing.T, cm *system.CleanupManager) *APIClient {
 	// blank config should result in using defaults in node.Node constructor
 	return setupNodeForTestWithConfig(t, cm, APIServerConfig{})
 }
 
-//nolint:unused // used in tests
 func setupNodeForTestWithConfig(t *testing.T, cm *system.CleanupManager, serverConfig APIServerConfig) *APIClient {
 	system.InitConfigForTesting(t)
 	ctx := context.Background()
 
 	libp2pPort, err := freeport.GetFreePort()
-	require.NoError(t, err)
-
-	apiPort, err := freeport.GetFreePort()
 	require.NoError(t, err)
 
 	libp2pHost, err := libp2p.NewHost(libp2pPort)
@@ -38,22 +33,18 @@ func setupNodeForTestWithConfig(t *testing.T, cm *system.CleanupManager, serverC
 	apiServer, err := NewAPIServer(APIServerParams{
 		Host:    libp2pHost,
 		Address: "0.0.0.0",
-		Port:    apiPort,
+		Port:    0,
 		Config:  serverConfig,
 	})
 	require.NoError(t, err)
 
-	go func() {
-		err := apiServer.ListenAndServe(ctx, cm)
-		require.NoError(t, err)
-	}()
+	require.NoError(t, apiServer.ListenAndServe(ctx, cm))
 
 	client := NewAPIClient(apiServer.GetURI())
 	require.NoError(t, waitForHealthy(ctx, client))
 	return client
 }
 
-//nolint:unused // used in tests
 func waitForHealthy(ctx context.Context, c *APIClient) error {
 	ch := make(chan bool)
 	go func() {

--- a/pkg/test/compute/setup_test.go
+++ b/pkg/test/compute/setup_test.go
@@ -60,12 +60,9 @@ func (s *ComputeSuite) setupNode() {
 	host, err := libp2p.NewHost(libp2pPort)
 	s.NoError(err)
 
-	apiPort, err := freeport.GetFreePort()
-	s.NoError(err)
-
 	apiServer, err := publicapi.NewAPIServer(publicapi.APIServerParams{
 		Address: "0.0.0.0",
-		Port:    apiPort,
+		Port:    0,
 		Host:    host,
 		Config:  publicapi.DefaultAPIServerConfig,
 	})

--- a/pkg/test/requester/publicapi/util.go
+++ b/pkg/test/requester/publicapi/util.go
@@ -36,9 +36,6 @@ func setupNodeForTestWithConfig(t *testing.T, config publicapi.APIServerConfig) 
 	libp2pPort, err := freeport.GetFreePort()
 	require.NoError(t, err)
 
-	apiPort, err := freeport.GetFreePort()
-	require.NoError(t, err)
-
 	libp2pHost, err := libp2p.NewHost(libp2pPort)
 	require.NoError(t, err)
 
@@ -46,7 +43,7 @@ func setupNodeForTestWithConfig(t *testing.T, config publicapi.APIServerConfig) 
 		CleanupManager:      system.NewCleanupManager(),
 		Host:                libp2pHost,
 		HostAddress:         "0.0.0.0",
-		APIPort:             apiPort,
+		APIPort:             0,
 		JobStore:            datastore,
 		ComputeConfig:       node.NewComputeConfigWithDefaults(),
 		RequesterNodeConfig: node.NewRequesterConfigWithDefaults(),

--- a/pkg/test/requester/publicapi/websocket_test.go
+++ b/pkg/test/requester/publicapi/websocket_test.go
@@ -4,8 +4,11 @@ package publicapi
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"net"
 	"testing"
+	"time"
 
 	"github.com/bacalhau-project/bacalhau/pkg/logger"
 	"github.com/bacalhau-project/bacalhau/pkg/model"
@@ -52,23 +55,33 @@ func (s *WebsocketSuite) TestWebsocketEverything() {
 
 	conn, _, err := websocket.DefaultDialer.Dial(url, nil)
 	require.NoError(s.T(), err)
+	s.T().Cleanup(func() {
+		s.NoError(conn.Close())
+	})
 
 	eventChan := make(chan model.JobEvent)
 	go func() {
+		defer close(eventChan)
 		for {
 			var event model.JobEvent
 			err = conn.ReadJSON(&event)
+			if errors.Is(err, net.ErrClosed) {
+				return
+			}
 			require.NoError(s.T(), err)
 			eventChan <- event
 		}
 	}()
+
+	// Pause to ensure the websocket connects _before_ we submit the job
+	time.Sleep(100 * time.Millisecond)
 
 	genericJob := testutils.MakeGenericJob()
 	_, err = s.client.Submit(ctx, genericJob)
 	require.NoError(s.T(), err)
 
 	event := <-eventChan
-	require.Equal(s.T(), "Created", event.EventName.String())
+	require.Equal(s.T(), model.JobEventCreated, event.EventName)
 
 }
 

--- a/pkg/util/multiaddresses/multiaddresses.go
+++ b/pkg/util/multiaddresses/multiaddresses.go
@@ -1,0 +1,32 @@
+package multiaddresses
+
+import (
+	"sort"
+
+	"github.com/multiformats/go-multiaddr"
+	"golang.org/x/exp/slices"
+)
+
+func SortLocalhostFirst(multiAddresses []multiaddr.Multiaddr) []multiaddr.Multiaddr {
+	multiAddresses = slices.Clone(multiAddresses)
+	preferLocalhost := func(m multiaddr.Multiaddr) int {
+		count := 0
+		if _, err := m.ValueForProtocol(multiaddr.P_TCP); err == nil {
+			count++
+		}
+		if ip, err := m.ValueForProtocol(multiaddr.P_IP4); err == nil {
+			count++
+			if ip == "127.0.0.1" {
+				count++
+			}
+		} else if ip, err := m.ValueForProtocol(multiaddr.P_IP6); err == nil && ip != "::1" {
+			count++
+		}
+		return count
+	}
+	sort.Slice(multiAddresses, func(i, j int) bool {
+		return preferLocalhost(multiAddresses[i]) > preferLocalhost(multiAddresses[j])
+	})
+
+	return multiAddresses
+}

